main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+        val negCstsSize = cstsNeg.size
+        val negVarsSize = varsNeg.size
+        val subformulas = elemPos.size + cstsPos.size + varsPos.size +
+          negElemSize + negCstsSize + negVarsSize + other.size
+        val connectives = negElemSize + negCstsSize + negVarsSize
+        counter += (subformulas - 1) + connectives
+        workList = other ++ workList
+      }
+
+      while (workList.nonEmpty) {
+        val t :: next = workList
+        workList = next
+        t match {
+          case Univ => ()
+          case Empty => ()
+          case Cst(_) => ()
+          case Var(_) => ()
+          case ElemSet(_) => ()
+          case Compl(t) =>
+            counter += 1
+            workList = t :: workList
+          case Inter(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+            countSetFormulas(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+          case Union(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+            countSetFormulas(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        }
+      }
+      counter
+    }
+
+    /** Returns a human-readable string of `this`. */
+    override final def toString: String = this match {
+      case Univ => "univ"
+      case Empty => "empty"
+      case Cst(c) => s"c$c"
+      case ElemSet(s) if s.sizeIs == 1 => s"e${s.head}"
+      case ElemSet(s) => s"e${s.mkString("+")}"
+      case Var(x) => s"x$x"
+      case Compl(f) => f match {
+        case Univ | Empty | Cst(_) | ElemSet(_) | Var(_) | Compl(_) => s"!$f"
+        case Inter(_, _, _, _, _, _, _) | Union(_, _, _, _, _, _, _) => s"!($f)"
+      }
+      case Inter(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+        val subformulas = subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        s"(${subformulas.mkString(" ∩ ")})"
+      case Union(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+        val subformulas = subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        s"(${subformulas.mkString(" ∪ ")})"
+    }
+
+  }
+
+  object SetFormula {
+
+    /** Skip invariant checks if `false`. */
+    private val CHECK_INVARIANTS: Boolean = true
+
+    /** A trait used to refer to [[Univ]] and [[Empty]] collectively. */
+    sealed trait UnivOrEmpty extends SetFormula
+
+    /** The full universe set (`univ`). */
+    final case object Univ extends SetFormula with UnivOrEmpty
+
+    /** The empty set (`empty`). */
+    final case object Empty extends SetFormula with UnivOrEmpty
+
+    /**
+      * An uninterpreted constant set (`c42`), i.e. a set that we do not know and cannot make
+      * assumptions about.
+      *
+      * Invariant: [[ElemSet]], [[Cst]], and [[Var]] must use disjoint integers.
+      */
+    final case class Cst(c: Int) extends SetFormula
+
+    /**
+      * A set variable (`x42`), i.e. a set that we do not know but should learn assumptions about.
```

Comment:
> but should learn assumptions about.

Perhaps something more like, a variable that can be instantiated by unification?